autolab/Autolab
# Implement Initial LTI 1.3 Launch

## Description
This PR implements the LTI Advantage launch flow described in [LMS Integration Design Doc V2.0](https://docs.google.com/document/d/1qbWM9Ll94QO70KlIav_URePKpAGCDn1NLktNSqjaKyo/edit).  It is recommended to view "Implementation and Technical Details" as well as "LTI Advantage Launch Flow" sections. The full LTI 1.3 launch specification for tools such as Autolab can be found [here](http://www.imsglobal.org/spec/security/v1p0/#step-1-third-party-initiated-login).

Most of the code in this PR is based off the [LTI 1.3 PHP library](https://github.com/IMSGlobal/lti-1-3-php-library/tree/master/src/lti) and the [pylti1.3 library](https://github.com/dmitry-viskov/pylti1.3), specifically [LTI_OIDC_Login.php](https://github.com/IMSGlobal/lti-1-3-php-library/blob/master/src/lti/LTI_OIDC_Login.php), and [LTI_Message_Launch.php](https://github.com/IMSGlobal/lti-1-3-php-library/blob/master/src/lti/LTI_Message_Launch.php).

Specific changes:
- Build initial LTI 1.3 Integration Library in `controllers/lti_launch_controller.rb`
  - `/lti_launch/oidc_login/` endpoint handles Step 1 and Step 2 of the launch flow (see design doc for more details),
    - Ensure that `iss` parameter from platform matches our configuration settings, `lti_message_hint` exists
    - create `state`, `nonce` params using `SecureRandom`, with `state` stored as a short lived cookie and `nonce` as a cache entry (to prevent replay attacks)
    - Generate authentication parameters based off rails configuration settings (in `lti_settings.yml`), and redirect to authorization url of platform.
   - `/lti_launch/launch/` endpoint handles Step 4 of the launch flow (see design doc for more details), 
     - mainly validation of data received from platform, and sends off received JWT body data to `lti_launch_initialize`
     - validate both `state` and `nonce` received match the ones generated during `oidc_login`
     - validate response from platform is a valid JWT with correct signature by platform
     - validate `client_id` received from platform to that in configuration settings
     - Ensure that request is a "LtiResourceLinkRequest" with version "1.3.0"
     - validate we have access to NRPS and that we receive "context_memberships_url"
- display final resource (course context, etc. in users/:id/lti_launch_initialize) via `lti_launch_initialize` in `users_controller.rb`

## How Has This Been Tested?



- Go to /config and run:
`cp lti_settings.yml.template lti_settings.yml`
- Fill in `lti_settings.yml`
  - iss: "https://lti-ri.imsglobal.org"
  - developer_key: "lti.app.client_id"
  - auth_url: "https://lti-ri.imsglobal.org/platforms/3644/authorizations/new"
  - Do not modify the public key, as the template already uses the testing platform's public key
  - (You can find these when viewing the testing platform overview, covered a couple steps down)
- enable cache by running `rails dev:cache` 
- Go to the  [LTI Reference Tool main page](https://lti-ri.imsglobal.org/)
- Go to "Manage Platforms"
- Click on "Unclaimed Platforms"
- Find the "Autolab Test" platform (ctrl-f)

- Click on "View Platform". Make sure client id, OIDC Auth URL, platform public key matches the rails config file.
<img width="1290" alt="Screen Shot 2022-11-08 at 15 20 57" src="https://user-images.githubusercontent.com/25730111/200667105-a1525dc5-0f17-4ee7-87ce-5e4057b1b239.png">

- Click on "Resource Links" for Autolab Test
- Click "Select User for Launch" for Resource Link with Title "Autolab PR test"

<img width="1287" alt="Screen Shot 2022-11-08 at 15 19 48" src="https://user-images.githubusercontent.com/25730111/200666907-d9ff58d2-7b8f-4ce4-bfc4-fe1cdce066fc.png">

- Choose user with "Student: No", and click "Launch Resource Link (OIDC)". For example, you should launch as Davida in the example below.
<img width="1213" alt="Screen Shot 2022-11-08 at 15 18 05" src="https://user-images.githubusercontent.com/25730111/200666627-2bc8726f-0c80-4f0e-b848-ecce7e221d73.png">

- Choose either to send a POST or GET request (either works) on the next page
- You should immediately be redirected to a page similar to the one below. If you have to log into Autolab, or receive a message that you are already logged into autolab, try going back to the Resource Link page and clicking "Launch Resource Link (OIDC)" again. If you received an error from the oidc_login endpoint, make sure your `lti_settings.yml` information is correct, and try again.

<img width="1276" alt="Screen Shot 2022-11-08 at 15 25 11" src="https://user-images.githubusercontent.com/25730111/200667978-f4b6e9e3-7853-4663-bb13-5e0dd8ba37b3.png">

- Click "launch resource link"
- Scroll down past the JWT and click "Perform Launch." Again, if you receive an auth request from Autolab, log in, but then restart the launch process. If you receive an error message along the lines of "state/nonce not correct" you may have been too slow to complete the launch flow. Otherwise, you should be viewing this page:

<img width="1106" alt="Screen Shot 2022-11-08 at 15 29 00" src="https://user-images.githubusercontent.com/25730111/200668709-13ae0f6a-1af0-4c71-9198-9bee20c3e2d8.png">

## Types of changes


## Checklist:

-> will update documentation in a later PR

To do: figure out why "already logged in" auth error shows up and interrupts launch flow
